File: src/agents/database-client.ts
/**
 * Direct Database Client for Agents
 * 
 * Agents can access the database directly without going through MCP protocol.
 * This is simpler for autonomous agents and avoids OAuth complexity.
 */

import postgres from "postgres";

export class AgentDatabaseClient {
	private db: postgres.Sql | null = null;
	private databaseUrl: string;

	constructor(databaseUrl: string) {
		this.databaseUrl = databaseUrl;
	}

	/**
	 * Connect to database
	 */
	async connect(): Promise<void> {
		if (!this.db) {
			this.db = postgres(this.databaseUrl, {
				max: 5,
				idle_timeout: 20,
				connect_timeout: 10,
			});
		}
	}

	/**
	 * Disconnect from database
	 */
	async disconnect(): Promise<void> {
		if (this.db) {
			await this.db.end();
			this.db = null;
		}
	}

	/**
	 * Get a random joke
	 */
	async getRandomJoke(category?: string, language?: string): Promise<any> {
		await this.connect();
		if (!this.db) throw new Error("Database not connected");

		let query = "SELECT * FROM jokes WHERE 1=1";
		const params: any[] = [];

		if (category) {
			query += " AND category = $1";
			params.push(category);
		}
		if (language) {
			query += ` AND language = $${params.length + 1}`;
			params.push(language);
		}

		query += " ORDER BY RANDOM() LIMIT 1";

		const result = await this.db.unsafe(query, params);
		return result.length > 0 ? result[0] : null;
	}

	/**
	 * Add a joke
	 */
	async addJoke(content: string, category?: string, language: string = "en", author?: string): Promise<any> {
		await this.connect();
		if (!this.db) throw new Error("Database not connected");

		const result = await this.db`
			INSERT INTO jokes (content, language, category, author)
			VALUES (${content}, ${language}, ${category || null}, ${author || "agent"})
			RETURNING *
		`;
		return result[0];
	}

	/**
	 * Rate a joke
	 */
	async rateJoke(jokeId: number, agentId: string, rating: number, comment?: string): Promise<any> {
		await this.connect();
		if (!this.db) throw new Error("Database not connected");

		const result = await this.db`
			INSERT INTO joke_ratings (joke_id, agent_id, rating, comment)
			VALUES (${jokeId}, ${agentId}, ${rating}, ${comment || null})
			ON CONFLICT (joke_id, agent_id)
			DO UPDATE SET rating = ${rating}, comment = ${comment || null}, created_at = CURRENT_TIMESTAMP
			RETURNING *
		`;
		return result[0];
	}

	/**
	 * Store agent memory
	 */
	async storeMemory(
		jokeId: number,
		agentId: string,
		rating: number,
		notes?: string,
		tags?: string[]
	): Promise<any> {
		await this.connect();
		if (!this.db) throw new Error("Database not connected");

		const result = await this.db`
			INSERT INTO agent_memories (agent_id, joke_id, rating, personal_notes, comedic_style_tags)
			VALUES (${agentId}, ${jokeId}, ${rating}, ${notes || null}, ${tags || null})
			ON CONFLICT (agent_id, joke_id)
			DO UPDATE SET
				rating = ${rating},
				personal_notes = ${notes || null},
				comedic_style_tags = ${tags || null},
				updated_at = CURRENT_TIMESTAMP
			RETURNING *
		`;
		return result[0];
	}

	/**
	 * Get joke ratings
	 */
	async getJokeRatings(jokeId: number): Promise<any> {
		await this.connect();
		if (!this.db) throw new Error("Database not connected");

		const ratings = await this.db`
			SELECT * FROM joke_ratings WHERE joke_id = ${jokeId} ORDER BY created_at DESC
		`;

		const avgResult = await this.db`
			SELECT AVG(rating) as avg_rating, COUNT(*) as rating_count
			FROM joke_ratings
			WHERE joke_id = ${jokeId}
		`;

		return {
			jokeId,
			averageRating: avgResult[0]?.avg_rating ? parseFloat(avgResult[0].avg_rating).toFixed(2) : null,
			ratingCount: parseInt(avgResult[0]?.rating_count || "0"),
			ratings,
		};
	}

	/**
	 * Search jokes
	 */
	async searchJokes(query?: string, category?: string, language?: string, limit: number = 10): Promise<any[]> {
		await this.connect();
		if (!this.db) throw new Error("Database not connected");

		let sql = "SELECT * FROM jokes WHERE 1=1";
		const params: any[] = [];
		let paramCount = 0;

		if (query) {
			paramCount++;
			sql += ` AND content ILIKE $${paramCount}`;
			params.push(`%${query}%`);
		}
		if (category) {
			paramCount++;
			sql += ` AND category = $${paramCount}`;
			params.push(category);
		}
		if (language) {
			paramCount++;
			sql += ` AND language = $${paramCount}`;
			params.push(language);
		}

		sql += ` ORDER BY created_at DESC LIMIT $${++paramCount}`;
		params.push(limit);

		return await this.db.unsafe(sql, params);
	}
}

File: pyagents/database_client.py
"""In-memory stand-in for the database client.

The TypeScript version connects to PostgreSQL. For the Python demo we keep
state in memory so the script can run without external services.
"""

import itertools
import random
from typing import Any, Dict, List, Optional


class AgentDatabaseClient:
	def __init__(self, database_url: Optional[str]):
		self.database_url = database_url
		self._jokes: List[Dict[str, Any]] = []
		self._ratings: List[Dict[str, Any]] = []
		self._ids = itertools.count(1)
		self._connected = False

	async def connect(self) -> None:
		self._connected = True
		if not self._jokes:
			self._seed_jokes()

	async def disconnect(self) -> None:
		self._connected = False

	async def get_random_joke(self, category: Optional[str] = None) -> Dict[str, Any]:
		filtered = [j for j in self._jokes if not category or j.get("category") == category]
		if not filtered:
			return {"id": None, "content": "No jokes available yet."}
		return random.choice(filtered)

	async def add_joke(
		self,
		content: str,
		category: Optional[str],
		language: str,
		agent_id: str,
	) -> Dict[str, Any]:
		joke = {
			"id": next(self._ids),
			"content": content,
			"category": category or "general",
			"language": language,
			"agent_id": agent_id,
		}
		self._jokes.append(joke)
		return joke

	async def rate_joke(
		self,
		joke_id: Optional[int],
		agent_id: str,
		rating: int,
		comment: Optional[str],
	) -> Dict[str, Any]:
		rating_entry = {
			"joke_id": joke_id,
			"agent_id": agent_id,
			"rating": rating,
			"comment": comment,
		}
		self._ratings.append(rating_entry)
		return rating_entry

	async def store_memory(
		self,
		joke_id: Optional[int],
		agent_id: str,
		rating: int,
		notes: Optional[str],
		tags: Optional[List[str]],
	) -> Dict[str, Any]:
		return {
			"joke_id": joke_id,
			"agent_id": agent_id,
			"rating": rating,
			"notes": notes,
			"tags": tags or [],
		}

	async def search_jokes(
		self,
		query: Optional[str],
		category: Optional[str],
		language: Optional[str],
		limit: int,
	) -> List[Dict[str, Any]]:
		matches = [
			j for j in self._jokes
			if (not query or query.lower() in j.get("content", "").lower())
			and (not category or j.get("category") == category)
			and (not language or j.get("language") == language)
		]
		return matches[:limit]

	def _seed_jokes(self) -> None:
		self._jokes.extend(
			[
				{
					"id": next(self._ids),
					"content": "Why did the agent cross the road? To get better context.",
					"category": "general",
					"language": "en",
					"agent_id": "seed",
				},
				{
					"id": next(self._ids),
					"content": "The database went on vacation. It needed a little replication.",
					"category": "tech",
					"language": "en",
					"agent_id": "seed",
				},
			]
		)
